src/peft/tuners/lora/conversion.py

```diff
@@ -0,0 +1,348 @@
+# Copyright 2025-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import os
+import pathlib
+import warnings
+
+import torch
+from safetensors.torch import save_file
+from tqdm import tqdm
+from transformers.pytorch_utils import Conv1D
+
+from peft.tuners.tuners_utils import BaseTunerLayer
+from peft.utils import SAFETENSORS_WEIGHTS_NAME
+from peft.utils.other import ModulesToSaveWrapper
+
+from .config import LoraConfig
+
+
+def _find_cutoff_index(S: torch.Tensor, threshold: float) -> int:
+    # assumes that the singular values are sorted
+    if S.dim() != 1:
+        raise ValueError("Input vector must be 1d.")
+
+    energy = S**2
+    cs = torch.cumsum(energy, dim=0)
+    total = cs[-1]
+
+    cutoff = threshold * total
+    # smallest index i with cs[i] >= cutoff
+    k = torch.searchsorted(cs, cutoff).item()
+    # k is the index of the first item that surpasses the threshold; since we want to include it, add + 1
+    return k + 1
+
+
+@torch.no_grad()
+def _convert_module_to_lora(
+    module: BaseTunerLayer, rank: int | float, adapter_name: str = "default"
+) -> tuple[torch.Tensor, torch.Tensor, int]:
+    """Convert a single BaseTunerLayer's adapter weight to a LoRA weight, return A, B, and the effective rank."""
+    if not hasattr(module, "get_delta_weight"):
+        # if we arrive here, it means that the layer actually does not support LoRA conversion, which should not happen
+        raise TypeError(
+            f"Module of type {type(module)} does not have a get_delta_weight method, which is required for conversion. "
+            "Please open an issue: https://github.com/huggingface/peft/issues"
+        )
```

Comment:
Would prefer to raise it from `convert_to_lora()`.

Reply:
Good point, moving it there